hedyorg/hedy
# 🖊️ Improved error message for missing colons in level 17

- Add preprocessing rules which take an argument in order to allow for a rule to turn into an error without copying the old rule definition. For example, `error_ifelse<old_rule_to_error ifelse>` is used in level 8 to transform the old flat if-else to an error.
- Add errors for if-else statements, loops and function definitions missing a colon in level 17

Fixes #5222

**How to test**
- Automated tests are added for the new preprocessing functionality and for the commands missing colons in level 17.
- To see the error message, run the following program in level 17 and try removing any of the colons at a time. The message should direct you to the right line.
```
define function with b:
    i = 1
    while i < b:
        print i
        i = i + 1

b = 10
if b is 1:
    call function with b
elif b is 2:
    call function with b
else:
    call function with b

if b is pressed:
    call function with b
elif c is pressed:
    call function with b
else:
    call function with b

```
